### PR TITLE
Added CORS headers to enable Front-End inputs on Go Engine

### PR DESCRIPTION
**Pull request by ericandrewmeadows**
# Description

* Added CORS headers to all requests
  * All front-end web requests were getting rejected
* Added OPTIONS Request type, which is done via CORS Preflight
  * All OPTIONS requests, typical in CORS Preflight, were being rejected

Fixes # [CORS issue in Slack](https://seldondev.slack.com/archives/C8Y9A8G0Y/p1591802505468300)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Unit Tests
- [ ] Integration Tests
- [x] Production deployment (at MLB)

**Test Configuration**:

- Go version: `go version go1.14.3 darwin/amd64`

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **axsaucedo**

If they are not already assigned, you can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Hi @ericandrewmeadows. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: executor/api/rest/middlewares.go [hunk 1]
```diff
@@ -1,9 +1,10 @@
 package rest
 
 import (
+	"net/http"
```
Comment (ericandrewmeadows) [inline executor/api/rest/middlewares.go:4]: This happened to just be part of my auto-save in VSCode - it ordered the imports here.

File: executor/api/rest/server.go [hunk 2]
```diff
@@ -127,30 +127,31 @@ func (r *SeldonRestApi) Initialise() {
...
 		switch r.Protocol {
 		case api.ProtocolSeldon:
 			//v0.1 API
-			api01 := r.Router.PathPrefix("/api/v0.1").Methods("POST").Subrouter()
+			api01 := r.Router.PathPrefix("/api/v0.1").Methods("OPTIONS", "POST").Subrouter()
 			api01.Handle("/predictions", r.wrapMetrics(metric.PredictionHttpServiceName, r.predictions))
 			api01.Handle("/feedback", r.wrapMetrics(metric.FeedbackHttpServiceName, r.feedback))
-			r.Router.NewRoute().Path("/api/v0.1/status/{" + ModelHttpPathVariable + "}").Methods("GET").HandlerFunc(r.wrapMetrics(metric.StatusHttpServiceName, r.status))
-			r.Router.NewRoute().Path("/api/v0.1/metadata/{" + ModelHttpPathVariable + "}").Methods("GET").HandlerFunc(r.wrapMetrics(metric.MetadataHttpServiceName, r.metadata))
+			r.Router.NewRoute().Path("/api/v0.1/status/{"+ModelHttpPathVariable+"}").Methods("GET", "OPTIONS").HandlerFunc(r.wrapMetrics(metric.StatusHttpServiceName, r.status))
```
Comment (ericandrewmeadows) [inline executor/api/rest/server.go:138]: OPTIONS, +Autoformatting

File: executor/api/rest/middlewares.go [hunk 3]
```diff
@@ -16,6 +17,13 @@ const (
 
 	contentTypeOptsHeader = "X-Content-Type-Options"
 	contentTypeOptsValue  = "nosniff"
+
+	corsAllowOriginHeader  = "Access-Control-Allow-Origin"
+	corsAllowOriginValue   = "*"
```
Comment (SachinVarghese) [inline executor/api/rest/middlewares.go:22]: The default values for these headers are very lax from the security perspective. Maybe these values should be set from a custom config and the defaults should be made more strict.

**Comment by seldondev** [comment]
Tue Jun 16 08:30:07 UTC 2020
The logs for [lint] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=3

**Comment by seldondev** [comment]
Tue Jun 16 08:30:20 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=1

**Comment by seldondev** [comment]
Tue Jun 16 08:30:20 UTC 2020
The logs for [integration] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=2

**Comment by seldondev** [comment]
@ericandrewmeadows: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 4140a115a9622bff407596d77c6f08f4f805663e | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@ericandrewmeadows: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 4140a115a9622bff407596d77c6f08f4f805663e | [link]() | `/test this`
integration | 4140a115a9622bff407596d77c6f08f4f805663e | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: executor/api/rest/middlewares.go (hunk 3 above)
Comment (ericandrewmeadows) [reply executor/api/rest/middlewares.go:22]: 👍 - I agree.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **adriangonz**

If they are not already assigned, you can assign the PR to them by writing `/assign @adriangonz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
@ericandrewmeadows: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 4140a115a9622bff407596d77c6f08f4f805663e | [link]() | `/test this`
integration | 4140a115a9622bff407596d77c6f08f4f805663e | [link]() | `/test integration`
lint | 65dfd7e6e13f983dd5b19b9f721826fcd54ed3b3 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@ericandrewmeadows: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 4140a115a9622bff407596d77c6f08f4f805663e | [link]() | `/test integration`
lint | 65dfd7e6e13f983dd5b19b9f721826fcd54ed3b3 | [link]() | `/test lint`
pr-build | 65dfd7e6e13f983dd5b19b9f721826fcd54ed3b3 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Tue Jun 16 18:42:46 UTC 2020
The logs for [lint] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=7

**Comment by seldondev** [comment]
Tue Jun 16 18:44:00 UTC 2020
The logs for [pr-build] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=6

**Comment by ericandrewmeadows** [comment]
@adriangonz - after diving in very deep on Gorilla's CORS implementation, it isn't actually setting the headers, even though the library looks as if it would, rendering it unusable.  This was tried a multitude of ways - I'm happy to chat through this on Slack tomorrow - But I went deep into testing and found that the library is just insufficient.

**Comment by ericandrewmeadows** [comment]
The only Gorilla method that works properly is `r.Router.Use(mux.CORSMethodMiddleware(r.Router))`, but this results in a return type that doesn't seem befitting of a quality web backend, so I would caution against using the Gorilla here as well because there should be only a single instance of each method in the return.  It works, but yields the following:
```
Access-Control-Allow-Methods: OPTIONS,POST,OPTIONS,POST,OPTIONS,POST
```
I have returned my code that handles a hard-coded list of the types because the request presentation won't confuse any devs going forward.

**Comment by adriangonz** [comment]
/test integration

Those are great insights on `gorilla/mux` @ericandrewmeadows! It's good to be aware of those limitations.

From my side, the changes look great now. I think this is ready to land once the integration tests pass.

**Comment by seldondev** [comment]
Wed Jun 17 09:09:34 UTC 2020
The logs for [integration] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=8

**Comment by seldondev** [comment]
@ericandrewmeadows: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 5d2a533b5ae1d8c25385f505f0cd5bd886c6e3e5 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Wed Jun 17 13:20:30 UTC 2020
The logs for [integration] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=9

**Comment by seldondev** [comment]
@ericandrewmeadows: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 5d2a533b5ae1d8c25385f505f0cd5bd886c6e3e5 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by adriangonz** [comment]
There are two failed tests, but TBH they just seem flaky (i.e. maybe waiting for Ambassador to come up). The core ones seem to pass.

- `TestBatchWorker.test_batch_worker`:
  https://github.com/SeldonIO/seldon-core/blob/9339683acb3092d4a8a75490d9aed0f2e0e44a37/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/9.log#L1617-L1668

- `TestPrepack.test_tfserving`:
  https://github.com/SeldonIO/seldon-core/blob/9339683acb3092d4a8a75490d9aed0f2e0e44a37/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/9.log#L1690-L1712

@axsaucedo do you think that the `test_batch_worker` one could be just due to flakiness?

**Comment by ericandrewmeadows** [comment]
@adriangonz - I was seeing those fail, and wasn't fully sure why.  We do intercept and add the headers to all requests, so I don't perceive that causing a failure here.

**Comment by axsaucedo** [comment]
Yes, it seems that we may need to increase the number of retries in the batch processor. 
/test integration

**Comment by seldondev** [comment]
Fri Jun 19 15:10:38 UTC 2020
The logs for [integration] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1952/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1952 --build=10

**Comment by adriangonz** [comment]
/approve

@axsaucedo thanks for clarifying that point! In that case, I think it's safe to land this one.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/1952#issuecomment-646714884" title="Approved">adriangonz</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [adriangonz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
@ericandrewmeadows: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 5d2a533b5ae1d8c25385f505f0cd5bd886c6e3e5 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by axsaucedo** [comment]
@adriangonz yes batch test passed, currently failed test is `TestTagsPythonS2iK8s.test_model_combiner_grpc` which i'm not sure why exactly is the case but it's not related to this at all. All other tests pass so shoudl eb good to go. I'll add an issue to increase retries on batch processor test

**Comment by seldondev** [comment]
failed to trigger Pull Request pipeline
* failed to create agent
* failed to calculate in repo config
* failed to load trigger config for repository SeldonIO/seldon-core for ref 5d2a533b5ae1d8c25385f505f0cd5bd886c6e3e5
* failed to find any lighthouse configuration files in repo SeldonIO/seldon-core at sha 5d2a533b5ae1d8c25385f505f0cd5bd886c6e3e5
* failed to process repo SeldonIO/seldon-core refref 5d2a533b5ae1d8c25385f505f0cd5bd886c6e3e5
* failed to list files in directory /var/tmp/gitrepo660600678/.lighthouse
* open /var/tmp/gitrepo660600678/.lighthouse
* no such file or directory

**Pull request merged**